# Aula 1 - Modelo de Regressão

## 1.1 Apresentação

## 1.2 Carregando os dados

[JSON Viewer](http://jsonviewer.stack.hu)

[Pandas: read_json](https://pandas.pydata.org/docs/reference/api/pandas.read_json.html?highlight=read_json#pandas.read_json)


In [ ]:
import pandas as pd
import numpy as np

In [ ]:
dados = pd.read_json(
    path_or_buf= '/content/sample_data/imoveis.json',
    orient='columns'
)

dados

ident  \
0      {'customerID': '775564-BOJSMVON', 'source': 'W...   
1      {'customerID': '660895-AUENKNYY', 'source': 'W...   
2      {'customerID': '751522-JESYFEQL', 'source': 'W...   
3      {'customerID': '714052-GAAEWYKS', 'source': 'W...   
4      {'customerID': '568886-ZIBFOMCC', 'source': 'W...   
...                                                  ...   
73610  {'customerID': '134148-MJIBCMDQ', 'source': 'W...   
73611  {'customerID': '541465-QBQVXXXZ', 'source': 'W...   
73612  {'customerID': '800542-BZREFTQS', 'source': 'W...   
73613  {'customerID': '939419-ECGUNTIH', 'source': 'W...   
73614  {'customerID': '610503-VULDJJZZ', 'source': 'W...   

                                                 listing  
0      {'types': {'unit': 'Outros', 'usage': 'Residen...  
1      {'types': {'unit': 'Apartamento', 'usage': 'Re...  
2      {'types': {'unit': 'Outros', 'usage': 'Residen...  
3      {'types': {'unit': 'Outros', 'usage': 'Comerci...  
4      {'types': {'unit': 'Apartamento', 'usage': 'Re...  
...                                                  ...  
73610  {'types': {'unit': 'Casa', 'usage': 'Residenci...  
73611  {'types': {'unit': 'Casa', 'usage': 'Residenci...  
73612  {'types': {'unit': 'Apartamento', 'usage': 'Re...  
73613  {'types': {'unit': 'Casa', 'usage': 'Residenci...  
73614  {'types': {'unit': 'Apartamento', 'usage': 'Re...  

[73615 rows x 2 columns]

### Explorar as informações

In [ ]:
dados.ident[0]

{'customerID': '775564-BOJSMVON', 'source': 'Website'}

In [ ]:
dados.listing[0]

{'address': {'city': 'Rio de Janeiro',
  'location': {'lat': -22.909429, 'lon': -43.413557},
  'neighborhood': 'Taquara',
  'zone': 'Zona Oeste'},
 'features': {'bathrooms': 0,
  'bedrooms': 0,
  'floors': 0,
  'parkingSpaces': 1,
  'suites': 0,
  'totalAreas': '62',
  'unitFloor': 0,
  'unitsOnTheFloor': 0,
  'usableAreas': '62'},
 'prices': {'price': '45000', 'tax': {'condo': '150', 'iptu': '0'}},
 'types': {'unit': 'Outros', 'usage': 'Residencial'}}

## 1.3 Tratamento de dados inicial

### Normalização

In [ ]:
dados_lista1 = pd.json_normalize(dados.ident)
dados_lista1.head()

customerID   source
0  775564-BOJSMVON  Website
1  660895-AUENKNYY  Website
2  751522-JESYFEQL  Website
3  714052-GAAEWYKS  Website
4  568886-ZIBFOMCC  Website

In [ ]:
dados_lista2 = pd.json_normalize(dados.listing, sep='_')
dados_lista2.head()

types_unit  types_usage    address_city  address_location_lon  \
0       Outros  Residencial  Rio de Janeiro            -43.413557   
1  Apartamento  Residencial  Rio de Janeiro            -43.509141   
2       Outros  Residencial  Rio de Janeiro            -43.646786   
3       Outros    Comercial  Rio de Janeiro            -43.330818   
4  Apartamento  Residencial  Rio de Janeiro            -43.480742   

   address_location_lat address_zone      address_neighborhood prices_price  \
0            -22.909429   Zona Oeste                   Taquara        45000   
1            -22.869698   Zona Oeste                Santíssimo        45000   
2            -22.986927   Zona Oeste        Pedra de Guaratiba        50000   
3            -22.881977   Zona Norte                 Cascadura        45000   
4            -23.027653   Zona Oeste  Recreio dos Bandeirantes        50000   

  prices_tax_iptu prices_tax_condo  features_bedrooms  features_bathrooms  \
0               0              150                  0                   0   
1               0              120                  2                   1   
2               0              100                  0                   0   
3             346              468                  0                   1   
4             120              400                  3                   2   

   features_suites  features_parkingSpaces features_usableAreas  \
0                0                       1                   62   
1                0                       1                   44   
2                0                       0                  132   
3                0                       0                   32   
4                1                       1                   60   

  features_totalAreas  features_floors  features_unitsOnTheFloor  \
0                  62                0                         0   
1                   0                0                         0   
2                 132                0                         0   
3                  32                0                         0   
4                   0                3                         4   

   features_unitFloor  
0                   0  
1                   0  
2                   0  
3                   3  
4                   2

### Junção dos dataframes

In [ ]:
dados_imoveis = pd.concat([dados_lista1, dados_lista2], axis=1)
dados_imoveis.head()

customerID   source   types_unit  types_usage    address_city  \
0  775564-BOJSMVON  Website       Outros  Residencial  Rio de Janeiro   
1  660895-AUENKNYY  Website  Apartamento  Residencial  Rio de Janeiro   
2  751522-JESYFEQL  Website       Outros  Residencial  Rio de Janeiro   
3  714052-GAAEWYKS  Website       Outros    Comercial  Rio de Janeiro   
4  568886-ZIBFOMCC  Website  Apartamento  Residencial  Rio de Janeiro   

   address_location_lon  address_location_lat address_zone  \
0            -43.413557            -22.909429   Zona Oeste   
1            -43.509141            -22.869698   Zona Oeste   
2            -43.646786            -22.986927   Zona Oeste   
3            -43.330818            -22.881977   Zona Norte   
4            -43.480742            -23.027653   Zona Oeste   

       address_neighborhood prices_price  ... prices_tax_condo  \
0                   Taquara        45000  ...              150   
1                Santíssimo        45000  ...              120   
2        Pedra de Guaratiba        50000  ...              100   
3                 Cascadura        45000  ...              468   
4  Recreio dos Bandeirantes        50000  ...              400   

  features_bedrooms  features_bathrooms  features_suites  \
0                 0                   0                0   
1                 2                   1                0   
2                 0                   0                0   
3                 0                   1                0   
4                 3                   2                1   

   features_parkingSpaces  features_usableAreas features_totalAreas  \
0                       1                    62                  62   
1                       1                    44                   0   
2                       0                   132                 132   
3                       0                    32                  32   
4                       1                    60                   0   

  features_floors  features_unitsOnTheFloor  features_unitFloor  
0               0                         0                   0  
1               0                         0                   0  
2               0                         0                   0  
3               0                         0                   3  
4               3                         4                   2  

[5 rows x 21 columns]

In [ ]:
dados_imoveis.shape

(73615, 21)

In [ ]:
for coluna in dados_imoveis.columns:
    print('========='*30)
    print(dados_imoveis[coluna].value_counts())

775564-BOJSMVON    1
917809-TRNXMOHE    1
125760-XEXDELYV    1
294183-VCUBOXCU    1
465524-CFDRZTHS    1
                  ..
200525-LDRZXCVM    1
938284-QWFBFUXB    1
882696-YVZSJYMI    1
566032-WFPGBXHL    1
610503-VULDJJZZ    1
Name: customerID, Length: 73615, dtype: int64
Website    73615
Name: source, dtype: int64
Apartamento    59106
Casa            9389
Outros          5120
Name: types_unit, dtype: int64
Residencial    69596
Comercial       4019
Name: types_usage, dtype: int64
Rio de Janeiro        73607
Queimados                 3
São Gonçalo               2
Cachoeirinha              1
São João de Meriti        1
Armação dos Búzios        1
Name: address_city, dtype: int64
-43.365895    399
-43.462590    278
-43.333066    236
-43.383834    232
-43.516473    218
             ... 
-43.231882      1
-43.236952      1
-43.240152      1
-43.185712      1
-43.250934      1
Name: address_location_lon, Length: 25753, dtype: int64
-23.000371    399
-23.017479    278
-22.937463    236
-2

## 1.4 Limpeza

### Filtro

In [ ]:
filtro = (dados_imoveis['types_usage'] == 'Residencial') & (dados_imoveis['address_city'] ==  'Rio de Janeiro')

#aplicando o filtro 
dados_imoveis = dados_imoveis[filtro]
dados_imoveis.head()

customerID   source   types_unit  types_usage    address_city  \
0  775564-BOJSMVON  Website       Outros  Residencial  Rio de Janeiro   
1  660895-AUENKNYY  Website  Apartamento  Residencial  Rio de Janeiro   
2  751522-JESYFEQL  Website       Outros  Residencial  Rio de Janeiro   
4  568886-ZIBFOMCC  Website  Apartamento  Residencial  Rio de Janeiro   
5  526755-OBLTYTEN  Website       Outros  Residencial  Rio de Janeiro   

   address_location_lon  address_location_lat address_zone  \
0            -43.413557            -22.909429   Zona Oeste   
1            -43.509141            -22.869698   Zona Oeste   
2            -43.646786            -22.986927   Zona Oeste   
4            -43.480742            -23.027653   Zona Oeste   
5            -43.571183            -22.966059   Zona Oeste   

       address_neighborhood prices_price  ... prices_tax_condo  \
0                   Taquara        45000  ...              150   
1                Santíssimo        45000  ...              120   
2        Pedra de Guaratiba        50000  ...              100   
4  Recreio dos Bandeirantes        50000  ...              400   
5                 Guaratiba        50000  ...                0   

  features_bedrooms  features_bathrooms  features_suites  \
0                 0                   0                0   
1                 2                   1                0   
2                 0                   0                0   
4                 3                   2                1   
5                 0                   0                0   

   features_parkingSpaces  features_usableAreas features_totalAreas  \
0                       1                    62                  62   
1                       1                    44                   0   
2                       0                   132                 132   
4                       1                    60                   0   
5                       0                   200                 200   

  features_floors  features_unitsOnTheFloor  features_unitFloor  
0               0                         0                   0  
1               0                         0                   0  
2               0                         0                   0  
4               3                         4                   2  
5               0                         0                   0  

[5 rows x 21 columns]

### Tipos de variáveis

In [ ]:
#atenção ao tipo das variáveis e o indice
dados_imoveis.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 69588 entries, 0 to 73614
Data columns (total 21 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   customerID                69588 non-null  object 
 1   source                    69588 non-null  object 
 2   types_unit                69588 non-null  object 
 3   types_usage               69588 non-null  object 
 4   address_city              69588 non-null  object 
 5   address_location_lon      69449 non-null  float64
 6   address_location_lat      69449 non-null  float64
 7   address_zone              69588 non-null  object 
 8   address_neighborhood      69588 non-null  object 
 9   prices_price              69588 non-null  object 
 10  prices_tax_iptu           60863 non-null  object 
 11  prices_tax_condo          63527 non-null  object 
 12  features_bedrooms         69588 non-null  int64  
 13  features_bathrooms        69588 non-null  int64  
 14  featur

Verificando a saída do `.info`

> 69588 entries, 0 to 73614

 percebemos que temos que resentar o index, para que as infos fiquem ordenadas corretamente.
 

In [ ]:
dados_imoveis.reset_index(drop=True, inplace=True)

In [ ]:
#mostrar o RangeIndex
dados_imoveis.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69588 entries, 0 to 69587
Columns: 21 entries, customerID to features_unitFloor
dtypes: float64(2), int64(7), object(12)
memory usage: 11.1+ MB


Agora está correto.

In [ ]:
dados_imoveis = dados_imoveis.astype({
    'prices_price': 'float64',
    'prices_tax_iptu': 'float64',
    'prices_tax_condo': 'float64',
    'features_usableAreas': 'int64',
    'features_totalAreas': 'int64',
})

dados_imoveis.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69588 entries, 0 to 69587
Data columns (total 21 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   customerID                69588 non-null  object 
 1   source                    69588 non-null  object 
 2   types_unit                69588 non-null  object 
 3   types_usage               69588 non-null  object 
 4   address_city              69588 non-null  object 
 5   address_location_lon      69449 non-null  float64
 6   address_location_lat      69449 non-null  float64
 7   address_zone              69588 non-null  object 
 8   address_neighborhood      69588 non-null  object 
 9   prices_price              69588 non-null  float64
 10  prices_tax_iptu           60863 non-null  float64
 11  prices_tax_condo          63527 non-null  float64
 12  features_bedrooms         69588 non-null  int64  
 13  features_bathrooms        69588 non-null  int64  
 14  featur

## 1.5 - Tratamento de dados final

### Lidando com dados nulos

#### Dados nulos zone

In [ ]:
dados_imoveis.address_zone.value_counts()

Zona Oeste      35523
Zona Sul        18535
Zona Norte      14379
Zona Central     1002
                  149
Name: address_zone, dtype: int64

In [ ]:
dados_imoveis['address_zone'] = dados_imoveis['address_zone'].replace('', np.nan)

In [ ]:
#verificação
dados_imoveis.address_zone.isnull().sum()

149

In [ ]:
#relação zona com bairro
dados_imoveis.head()

customerID   source   types_unit  types_usage    address_city  \
0  775564-BOJSMVON  Website       Outros  Residencial  Rio de Janeiro   
1  660895-AUENKNYY  Website  Apartamento  Residencial  Rio de Janeiro   
2  751522-JESYFEQL  Website       Outros  Residencial  Rio de Janeiro   
3  568886-ZIBFOMCC  Website  Apartamento  Residencial  Rio de Janeiro   
4  526755-OBLTYTEN  Website       Outros  Residencial  Rio de Janeiro   

   address_location_lon  address_location_lat address_zone  \
0            -43.413557            -22.909429   Zona Oeste   
1            -43.509141            -22.869698   Zona Oeste   
2            -43.646786            -22.986927   Zona Oeste   
3            -43.480742            -23.027653   Zona Oeste   
4            -43.571183            -22.966059   Zona Oeste   

       address_neighborhood  prices_price  ...  prices_tax_condo  \
0                   Taquara       45000.0  ...             150.0   
1                Santíssimo       45000.0  ...             120.0   
2        Pedra de Guaratiba       50000.0  ...             100.0   
3  Recreio dos Bandeirantes       50000.0  ...             400.0   
4                 Guaratiba       50000.0  ...               0.0   

   features_bedrooms  features_bathrooms  features_suites  \
0                  0                   0                0   
1                  2                   1                0   
2                  0                   0                0   
3                  3                   2                1   
4                  0                   0                0   

   features_parkingSpaces  features_usableAreas  features_totalAreas  \
0                       1                    62                   62   
1                       1                    44                    0   
2                       0                   132                  132   
3                       1                    60                    0   
4                       0                   200                  200   

   features_floors  features_unitsOnTheFloor  features_unitFloor  
0                0                         0                   0  
1                0                         0                   0  
2                0                         0                   0  
3                3                         4                   2  
4                0                         0                   0  

[5 rows x 21 columns]

In [ ]:
dici = dados_imoveis[~dados_imoveis['address_zone'].isna()].drop_duplicates(subset=['address_neighborhood']).to_dict('records')
print(dici)

[{'customerID': '775564-BOJSMVON', 'source': 'Website', 'types_unit': 'Outros', 'types_usage': 'Residencial', 'address_city': 'Rio de Janeiro', 'address_location_lon': -43.413557, 'address_location_lat': -22.909429, 'address_zone': 'Zona Oeste', 'address_neighborhood': 'Taquara', 'prices_price': 45000.0, 'prices_tax_iptu': 0.0, 'prices_tax_condo': 150.0, 'features_bedrooms': 0, 'features_bathrooms': 0, 'features_suites': 0, 'features_parkingSpaces': 1, 'features_usableAreas': 62, 'features_totalAreas': 62, 'features_floors': 0, 'features_unitsOnTheFloor': 0, 'features_unitFloor': 0}, {'customerID': '660895-AUENKNYY', 'source': 'Website', 'types_unit': 'Apartamento', 'types_usage': 'Residencial', 'address_city': 'Rio de Janeiro', 'address_location_lon': -43.509141, 'address_location_lat': -22.869698, 'address_zone': 'Zona Oeste', 'address_neighborhood': 'Santíssimo', 'prices_price': 45000.0, 'prices_tax_iptu': 0.0, 'prices_tax_condo': 120.0, 'features_bedrooms': 2, 'features_bathrooms':

In [ ]:
#dicionario da bairro para zona
dic_zonas = {dic['address_neighborhood']: dic['address_zone'] for dic in dici}
print(dic_zonas)

{'Taquara': 'Zona Oeste', 'Santíssimo': 'Zona Oeste', 'Pedra de Guaratiba': 'Zona Oeste', 'Recreio dos Bandeirantes': 'Zona Oeste', 'Guaratiba': 'Zona Oeste', 'Jacarezinho': 'Zona Norte', 'Santa Cruz': 'Zona Oeste', 'Vargem Grande': 'Zona Oeste', 'Bangu': 'Zona Oeste', 'Tanque': 'Zona Oeste', 'Santo Cristo': 'Zona Central', 'Coelho Neto': 'Zona Norte', 'Bento Ribeiro': 'Zona Norte', 'Campo Grande': 'Zona Oeste', 'Oswaldo Cruz': 'Zona Norte', 'Camorim': 'Zona Oeste', 'Sepetiba': 'Zona Oeste', 'Vigário Geral': 'Zona Norte', 'Campinho': 'Zona Norte', 'Realengo': 'Zona Oeste', 'Jacarepaguá': 'Zona Oeste', 'Cidade de Deus': 'Zona Oeste', 'Honório Gurgel': 'Zona Norte', 'Centro': 'Zona Central', 'Cosmos': 'Zona Oeste', 'Tomás Coelho': 'Zona Norte', 'Cascadura': 'Zona Norte', 'Pavuna': 'Zona Norte', 'Paciência': 'Zona Oeste', 'Praça Seca': 'Zona Oeste', 'Vila Valqueire': 'Zona Oeste', 'Inhoaíba': 'Zona Oeste', 'Vicente de Carvalho': 'Zona Norte', 'Senador Vasconcelos': 'Zona Oeste', 'Turiaçu'

In [ ]:
for bairro,zona in dic_zonas.items(): dados_imoveis.loc[dados_imoveis['address_neighborhood'] == bairro, 'address_zone'] = zona

In [ ]:
dados_imoveis['address_zone'].isnull().sum()

0

#### Dados nulos prices

In [ ]:
dados_imoveis.prices_tax_iptu.isnull().sum()

8725

In [ ]:
dados_imoveis.prices_tax_condo.isnull().sum()

6061

In [ ]:
dados_imoveis.prices_tax_iptu.fillna(0, inplace=True)
dados_imoveis.prices_tax_condo.fillna(0, inplace=True)

In [ ]:
print(f"Total de prices nulos: {dados_imoveis.prices_tax_iptu.isnull().sum()}")
print(f"Total de prices_tax_condo nulos: {dados_imoveis.prices_tax_condo.isnull().sum()}")

Total de prices nulos: 0
Total de prices_tax_condo nulos: 0


### Alterar o nome das colunas e seleção de variáveis

In [ ]:
dados_imoveis.drop(['customerID', 'source', 'types_usage', 'address_city', 
                    'address_location_lon', 'address_location_lat','address_neighborhood' ], axis = 1, inplace=True)

In [ ]:
dicionario_colunas = {'types_unit': 'unit', 'address_zone': 'zone', 'prices_price': 'price',
                     'prices_tax_condo': 'tax_condo', 'prices_tax_iptu': 'tax_iptu', 'features_bedrooms': 'bedrooms',
                      'features_bathrooms': 'bathrooms', 'features_suites': 'suites', 'features_parkingSpaces': 'parkingSpaces',
                      'features_usableAreas': 'usableAreas', 'features_totalAreas':'totalAreas', 'features_floors':'floors',
                      'features_unitsOnTheFloor':'unitsOnTheFloor', 'features_unitFloor':'unitFloor'}

dados_imoveis = dados_imoveis.rename(dicionario_colunas, axis = 1)
dados_imoveis.head()

unit        zone    price  tax_iptu  tax_condo  bedrooms  bathrooms  \
0       Outros  Zona Oeste  45000.0       0.0      150.0         0          0   
1  Apartamento  Zona Oeste  45000.0       0.0      120.0         2          1   
2       Outros  Zona Oeste  50000.0       0.0      100.0         0          0   
3  Apartamento  Zona Oeste  50000.0     120.0      400.0         3          2   
4       Outros  Zona Oeste  50000.0       0.0        0.0         0          0   

   suites  parkingSpaces  usableAreas  totalAreas  floors  unitsOnTheFloor  \
0       0              1           62          62       0                0   
1       0              1           44           0       0                0   
2       0              0          132         132       0                0   
3       1              1           60           0       3                4   
4       0              0          200         200       0                0   

   unitFloor  
0          0  
1          0  
2          0  
3          2  
4          0

# Aula 2 - Regressão linear

## 2.1 - Correlação

*   Correlação: grau da relação entre variáveis. Pode ser positiva ou negativa essa relação. 

*   Multicolinearidade: relação perfeita ou exata entre variáveis.



**Dica de leitura:** [Estatística prática para cientistas de dados: 50 conceitos essenciais](https://www.amazon.com.br/Estat%C3%ADstica-Pr%C3%A1tica-Para-Cientistas-Dados/dp/855080603X/ref=sr_1_21?__mk_pt_BR=%C3%85M%C3%85%C5%BD%C3%95%C3%91&crid=1JJ44C5LJXLR3&keywords=data+science&qid=1657892785&sprefix=data+science%2Caps%2C201&sr=8-21)
e [Estatística: O que é, para que serve e como funciona](https://www.amazon.com.br/Estat%C3%ADstica-para-serve-como-funciona/dp/8537815128/ref=pd_sbs_sccl_3_1/137-9319794-7502061?pd_rd_w=EQoo7&content-id=amzn1.sym.d27ac626-cb3c-4c30-830f-0fd405f2510f&pf_rd_p=d27ac626-cb3c-4c30-830f-0fd405f2510f&pf_rd_r=E896BYW6W206YVVR3SHA&pd_rd_wg=3O0Cg&pd_rd_r=3e3910b0-5f34-4a57-87b2-1d918494d614&pd_rd_i=8537815128&psc=1) [(Naked Statistics: Stripping the Dread from the Data - versão em inglês)](https://www.amazon.com.br/Naked-Statistics-Stripping-Dread-Data/dp/039334777X/ref=sr_1_2?keywords=naked+statistics&qid=1657893464&sprefix=naked+st%2Caps%2C267&sr=8-2&ufe=app_do%3Aamzn1.fos.6a09f7ec-d911-4889-ad70-de8dd83c8a74)

Para mais detalhes consultar [Documentação Seaborn](https://seaborn.pydata.org/generated/seaborn.heatmap.html)

Análise da distribuição da variável resposta.

### Transformação das variáveis 

Estamos realizando uma transformação logarítmica das variáveis para obtermos uma distribuição simétrica (mais próxima da normal), para que possamos usar o modelo de regressão linear. A transformação realizada é:


$$log1p = log(1+x)$$


Lembrar de realizar a transformação inversa para obter o valor original:


$$expm1 = exp(x) - 1$$ 




* Para mais detalhes consultar as documentações [`log1p`](https://numpy.org/doc/stable/reference/generated/numpy.log1p.html), [`expm1`](https://numpy.org/doc/stable/reference/generated/numpy.expm1.html#numpy.expm1) e [`FunctionTransformer`](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.FunctionTransformer.html#sklearn.preprocessing.FunctionTransformer)



### Dummies

## 2.2 - Regressão Linear

* "A análise de regressão diz respeito ao estudo da dependência de uma variável, a variável dependente, em relação a uma ou mais variáveis, as variáveis explanatórias, visando estimar e/ou prever o valor médio (da população) da primeira em termos dos valores conhecidos ou fixados (em amostragens repetidas) das segundas."

* [Documentação Sklearn - LinearRegression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html#sklearn.linear_model.LinearRegression) 

* Regressão Simples: a variável dependente (regressando ou `y`) se relaciona a uma única variável explanatória (regressor ou `X`).

* Regressão Múltipla: a variável dependente (regressando ou `y`) se relaciona a mais de uma única variável explanatória (regressor ou `Xi`).


[slides]

## 2.3 - Ajuste e Previsão

### Divisão dos dados

### Modelo

## 2.4 - Métrica de desempenho

### Coeficiente de Determinação 

* descobrir quão “bem” uma linha de regressão amostral é adequada aos dados
* r2 situa-se entre 0 e 1
* "r2 mede a proporção ou percentual da variação total de Y explicada pelo modelo de regressão."





### Extra

* [Documentação Sklearn - Métricas para Regressão](https://scikit-learn.org/stable/modules/model_evaluation.html#regression-metrics)

* Observação: Estas métricas dependem da escala dos dados!!!

* MSE (Mean Squared Error ou Erro Quadrático Médio): média da diferença **elevada ao quadrado** entre o valor real e o previsto. *(penalidade sobre o erro)*

* MAE (Mean Absolute Error ou Erro Absoluto Médio): média da diferença **absoluta** (módulo) entre o valor real e o previsto.


# Aula 3 - Árvore de Decisão para Regressão






## 3.1 - Árvore de Decisão Classificação versus Regressão

[slides]

[Biblioteca Sklearn - Desicion Tree](https://scikit-learn.org/stable/modules/tree.html#regression)

**Dica de leitura**: [Capítulo 6 - Mãos à obra: aprendizado de máquina com Scikit-Learn, Keras & TensorFlow: Conceitos, ferramentas e técnicas para a construção de sistemas inteligentes](https://www.amazon.com.br/M%C3%A3os-obra-aprendizado-Scikit-Learn-inteligentes/dp/8550815489/ref=sr_1_2?keywords=machine+learning&qid=1658181249&sprefix=machine%2Caps%2C469&sr=8-2&ufe=app_do%3Aamzn1.fos.6d798eae-cadf-45de-946a-f477d47705b9), [Para saber mais: como funciona a árvore de decisão? - Classificação](https://cursos.alura.com.br/course/machine-learning-classificacao-tras-panos/task/107966)



## 3.2 - Implementando o Modelo


**Gráfico de erro de predição**: Um gráfico de erros de predição mostra os alvos reais em relação aos valores previstos. Em um modelo perfeito, esses pontos estariam alinhados em 45 grau.


Para saber mais sobre outras visualizações:
[**Documentação YellowBrick**](https://www.scikit-yb.org/en/latest/)



## 3.3 - Métrica

# Aula 4 - Random Forest para Regressão


## 4.1 - Modelo Ensemble

[slides]

[Biblioteca Sklearn - Random Forest Regressor](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html#sklearn.ensemble.RandomForestRegressor)

**Dica de leitura**: [Artigo Científico - Random Forests](https://link.springer.com/article/10.1023/A:1010933404324), [Wikipédia: Random Forests](https://en.wikipedia.org/wiki/Random_forest#:~:text=Random%20forests%20or%20random%20decision,class%20selected%20by%20most%20trees.), [Capítulo 4 - Livro Introduction to Statistical and Machine Learning Methods for Data Science](https://www.amazon.com.br/dp/B09C2DVZHR/ref=dp-kindle-redirect?_encoding=UTF8&btkr=1), [Capítulo 7 - Mãos à obra: aprendizado de máquina com Scikit-Learn, Keras & TensorFlow: Conceitos, ferramentas e técnicas para a construção de sistemas inteligentes](https://www.amazon.com.br/M%C3%A3os-obra-aprendizado-Scikit-Learn-inteligentes/dp/8550815489/ref=sr_1_2?keywords=machine+learning&qid=1658181249&sprefix=machine%2Caps%2C469&sr=8-2&ufe=app_do%3Aamzn1.fos.6d798eae-cadf-45de-946a-f477d47705b9) 

## 4.2 - Previsão

**Ponto de Atenção**
Os parâmetros são praticamente os mesmos entre o `DecisionTreeRegressor()` e o `RandomForestRegressor()`, com exceção de:

* `n_estimators`: quantidade de árvores na floresta
* `bootstrap`: define se será utilizado o método estatístico de amostragem *bootstrap*
* `oob_score`: define se será utilizado o método de amostragem *out-of-bag* para mensurar os erros da predição
* `n_jobs`: número de *jobs* rodando paralelamente (computação paralela - processadores)
* `warm_start`: faz a adequação de uma nova floresta ou usa uma floresta existente.
* `max_samples`: para o caso do `bootstrap` ser `True`, define o número de amostras a serem extraídas de X para treinar cada árvore base. 

**Gráfico de erro de predição**

## 4.3 - Métrica de desempenho

# Aula 5 - Técnicas de Otimização


## 5.1 - Ferramentas de Otimização

* Por que otimizar?

* Quais ferramentas temos?

[Melhorando os hiperparâmetros - Documentação Sklearn](https://scikit-learn.org/stable/modules/grid_search.html#tuning-the-hyper-parameters-of-an-estimator)

### 1. Cross Validation (Validação Cruzada)

* [Documentação Cross-validation](https://scikit-learn.org/stable/modules/cross_validation.html)

* **Cross Validation** é uma técnica muito utilizada para avaliação de desempenho de modelos e consequente verificação de tendências no aprendizado de máquina. Ele consiste em particionar os dados em subconjuntos (partes), onde um subconjunto é utilizado para treino e outro subconjunto é utilizado para teste e avaliação do desempenho do modelo.

* Os dados particionados são os de treino (70% dos dados originais)!

* **K-fold**: conjunto de treino é dividido de forma aleatória em K subconjuntos de aproximadamente mesmo tamanho, depois são realizados K iterações, onde K-1 subconjuntos são para treino e 1 para validação.

* O desempenho calculado pela validação cruzada k-fold é a média de todos os valores calculados em cada iteração.


### 2. Randomized Parameter Optimization (Otimização Aleatória de Parâmetros)

* [Documentação Randomized Search](https://scikit-learn.org/stable/modules/grid_search.html#randomized-parameter-optimization)

* [Artigo Científico sobre Random Search](https://www.jmlr.org/papers/volume13/bergstra12a/bergstra12a.pdf)

* "Define um espaço de busca como um domínio limitado de valores de hiperparâmetros e pontos de amostra aleatórios nesse domínio."

* Imagem no artigo sobre funcionamento desta busca



## 5.2 - Otimizando a árvore de decisão

### Cross validation


### Random Search Validação

### Random Search Teste

## 5.3 - Otimizando a Random Forest

### Cross validation

### Random Search Validação

Definindo os parâmetros

### Random Search Teste

## 5.4 - Escolhendo o melhor modelo